llvm/lib/Target/AArch64/AArch64Processors.td

```diff
@@ -812,178 +812,270 @@ def ProcessorFeatures {
-                     [TuneX2]>;
-def : ProcessorModel<"cortex-x3", NeoverseN2Model, ProcessorFeatures.X3,
-                     [TuneX3]>;
-def : ProcessorModel<"cortex-x4", NeoverseN2Model, ProcessorFeatures.X4,
-                     [TuneX4]>;
-def : ProcessorModel<"neoverse-e1", CortexA53Model,
-                     ProcessorFeatures.NeoverseE1, [TuneNeoverseE1]>;
-def : ProcessorModel<"neoverse-n1", NeoverseN1Model,
-                     ProcessorFeatures.NeoverseN1, [TuneNeoverseN1]>;
-def : ProcessorModel<"neoverse-n2", NeoverseN2Model,
-                     ProcessorFeatures.NeoverseN2, [TuneNeoverseN2]>;
-def : ProcessorModel<"neoverse-n3", NeoverseN2Model,
-                     ProcessorFeatures.NeoverseN3, [TuneNeoverseN3]>;
-def : ProcessorModel<"neoverse-512tvb", NeoverseV1Model,
-                     ProcessorFeatures.Neoverse512TVB, [TuneNeoverse512TVB]>;
-def : ProcessorModel<"neoverse-v1", NeoverseV1Model,
-                     ProcessorFeatures.NeoverseV1, [TuneNeoverseV1]>;
-def : ProcessorModel<"neoverse-v2", NeoverseV2Model,
-                     ProcessorFeatures.NeoverseV2, [TuneNeoverseV2]>;
-def : ProcessorModel<"neoverse-v3", NeoverseV2Model,
-                     ProcessorFeatures.NeoverseV3, [TuneNeoverseV3]>;
-def : ProcessorModel<"neoverse-v3ae", NeoverseV2Model,
-                     ProcessorFeatures.NeoverseV3AE, [TuneNeoverseV3AE]>;
-def : ProcessorModel<"exynos-m3", ExynosM3Model, ProcessorFeatures.ExynosM3,
-                     [TuneExynosM3]>;
-def : ProcessorModel<"exynos-m4", ExynosM4Model, ProcessorFeatures.ExynosM4,
-                     [TuneExynosM4]>;
-def : ProcessorModel<"exynos-m5", ExynosM5Model, ProcessorFeatures.ExynosM4,
-                     [TuneExynosM4]>;
-def : ProcessorModel<"falkor", FalkorModel, ProcessorFeatures.Falkor,
-                     [TuneFalkor]>;
-def : ProcessorModel<"saphira", FalkorModel, ProcessorFeatures.Saphira,
-                     [TuneSaphira]>;
-def : ProcessorModel<"kryo", KryoModel, ProcessorFeatures.A53, [TuneKryo]>;
+
+// Alias for the latest Apple processor model supported by LLVM.
+def : ProcessorModel<"apple-latest", CycloneModel, ProcessorFeatures.AppleA16,
+                     [TuneAppleA16]>;
+
+def : AArch64Processor<"cortex-a34", HasV8_0aOps, CortexA53Model, ProcessorFeatures.A53,
+                     [TuneA35],
+                     [FeatureAES, FeatureSHA2, FeatureCRC]>;
+def : AArch64Processor<"cortex-a35", HasV8_0aOps, CortexA53Model, ProcessorFeatures.A53,
+                     [TuneA35],
+                     [FeatureAES, FeatureSHA2, FeatureCRC]>;
+def : AArch64Processor<"cortex-a53", HasV8_0aOps, CortexA53Model, ProcessorFeatures.A53,
+                     [TuneA53],
+                     [FeatureAES, FeatureSHA2, FeatureCRC]>;
+def : AArch64Processor<"cortex-a55", HasV8_2aOps, CortexA55Model, ProcessorFeatures.A55,
+                     [TuneA55],
+                     [FeatureAES, FeatureSHA2, FeatureFullFP16, FeatureDotProd, FeatureRCPC]>;
```

Comment:
As I said in the other review, why do we need to differentiate [FeatureAES, FeatureSHA2, FeatureFullFP16, FeatureDotProd, FeatureRCPC] from ProcessorFeatures.A55? If they differ it seems like a (historical) bug.

Reply:
And for that matter, why do we need `HasV8_2Ops` here when it's already in `ProcessorFeatures.A55`?

Reply:
The answer having two separate lists for the Architecture definitions is that one is a mandatory list and the other is the "enabled by default" list. I kept the same structure here but I am less confident that it is needed, I will double check and merge the two lists if possible.

`HasV8_2Ops` is added so that there is a top-level reference to the architecture for this processor. In `ProcessorFeatures.A55` it serves to add all of the implied features of `HasV8_2Ops` to the processor. It is true there is some redundancy here, but these are the options I see:
- Keep this redundancy, and add a tablegen check to make sure they are always consistent.
- Use the first feature in the feature list as the architecture for this processor (and assert that the first feature is an architecture etc in tablegen). This removes the redundancy but it seems a bit "magic".
- Omit `HasV8_2Ops` from `ProcessorFeatures.A55` and then somehow programatically add the architecture and its implied features when expanding the features for the cpu. I can look into this but it seemed like it was going to be more obscure.

Do you have a preference or another suggestions?

Reply:
This looks much better.

Reply:
I still don't follow why `HasV8_2Ops` needs to be an entry on `AArch64Processor` instead of it always living on the `ProcessorFeatures` record. What do we gain from having it there?

Reply:
It does sound a little better, though I must admit I am partially motivated by having downstream diff in this area. Churn here, and in TargetParserTest.cpp can be kinda painful, so I want to make sure that pain is for a good cause.